Valid and Same padding option in convolution layers.

walragatver opened the pull request:
Hi,
I have modified `convolutional` layer. But I have a doubt. Should we take all four padding dimension's as an argument in `convolution.hpp`. Currently we only take two dimension as argument in `constructor`. I have kept it as it is for now. I think taking all four padding dimension's would be good but it may break backward compatibility. Let me know what you think.

zoq commented on src/mlpack/methods/ann/layer/convolution.hpp:
```diff
@@ -74,7 +75,8 @@ class Convolution
               const size_t inputWidth = 0,
-              const size_t inputHeight = 0);
+              const size_t inputHeight = 0,
+              const std::string paddingType = "None");
```
I'm wondering if implementing another template argument (policy) would be more consistent with the existing mlpack code, let me know what you think. I don't have a preference here.

walragatver replied:
It looks like we cannot use `string literals` as template arguments. Should I create new classes for policies like we did for `GAN` to have `StandardGan` and `DCGAN` as class?

walragatver replied:
Currently I am keeping it same. It looks like I will need to create classes for it as a template argument. Let me know if it's incorrect. I will change it.

zoq replied:
I think a string is fine, even closer to tf/keras.

zoq commented on src/mlpack/methods/ann/layer/convolution_impl.hpp:
```diff
@@ -118,10 +133,10 @@ void Convolution<
   {
-    inputPaddedTemp.set_size(inputTemp.n_rows + padW * 2,
-        inputTemp.n_cols + padH * 2, inputTemp.n_slices);
+    inputPaddedTemp.set_size(inputTemp.n_rows + padWLeft + padWRight,
```
Would be great if we could add a test as well.

walragatver replied:
Actually I am doing this for `Inception Layer`, so I think we will know whether it is wrong or not. Logically It seems correct. Let me know what you think.

zoq replied:
Sorry for the slow response, ideally we can write an independent test for the adjusted conv operation, similar to https://github.com/mlpack/mlpack/blob/444559d71e1d29ef518c836aa571b40850a8755c/src/mlpack/tests/convolution_test.cpp#L124.

zoq commented:
We could implement another constructor that takes the other dimensions as well, that way we could be backward compatible.

zoq commented on src/mlpack/methods/ann/layer/convolution_impl.hpp:
```diff
@@ -92,6 +92,62 @@ Convolution<
+    outputHeight(0)
+{
+  weights.set_size((outSize * inSize * kW * kH) + outSize, 1);
+  if (paddingType == "Valid")
```
Perhaps we should transform ``paddingType`` to lower case and compare with ``"valid"`` what do you think?

walragatver replied:
Yes done.

walragatver commented:
Hi @zoq, I will add test soon. Sorry for the delay.

walragatver commented:
Hi @zoq,
I have added the test. I am thinking to work on `transposed convolution` in a separate PR. I have added tests in `ann_layer_test.cpp` file. Let me know if it's not the correct position.

zoq commented:
Okay, will take a look, thanks.

walragatver commented:
Also, 
Can I add this to the next milestone as @sreenikSS, pointed out its a necessity. I think only a review is required for this as well.

zoq commented on src/mlpack/tests/ann_layer_test.cpp:
```diff
@@ -2729,4 +2770,42 @@ BOOST_AUTO_TEST_CASE(LayerNormSerializationTest)
+  input = arma::linspace<arma::colvec>(0, 48, 49);
+  module1.Parameters() = arma::mat(9 + 1, 1, arma::fill::zeros);
+  module1.Reset();
+  module1.Forward(std::move(input), std::move(output));
```
Would be a good idea to check the output as well, but if you are going to test this is combination with the  Inception layer this is fine as well.

walragatver replied:
Yes, you are correct. I am now checking the `sum of all values of output as zero`. Let me know if anything else is needed to be added.

walragatver commented:
Hi @zoq,
It looks like the build is failing because of this. The build was passing in my machine as I was using C++14. What should we do now? Can we upgrade our `C++` version while building? I am not sure what I am saying let me know what you think.
https://stackoverflow.com/questions/32084706/why-initialization-doesnt-work-for-tuple

rcurtin commented:
@walragatver it's a little bit uglier, but you could do `std::tuple<size_t, size_t>(1, 1)` and that should work. :+1:  Given how old some of the compilers are for still-in-use distros like RHEL6, I don't think that we should change to require C++14 quite yet.
